index.bs

```diff
@@ -931,6 +1019,231 @@ steps:
+
+   Note: Implementations are expected to behave as if the skipped over file contents
+   are indeed filled with NUL bytes. That doesn't mean these bytes have to actually be
+   written to disk and take up disk space. Instead most file systems support so called
+   sparse files, where these NUL bytes don't take up actual disk space.
+
+1. Let |head| be a [=byte sequence=] containing the first |writePosition| bytes of |fileContents|.
+1. Let |tail| be an empty [=byte sequence=].
+1. If |writePosition| + |data|'s [=byte sequence/length=] is smaller than |oldSize|:
+   1. Set |tail| to a [=byte sequence=] containing the last
+      |oldSize| &minus; (|writePosition| + |data|'s [=byte sequence/length=]) bytes of |fileContents|.
+1. Let |newSize| be |head|'s [=byte sequence/length=] + |data|'s [=byte sequence/length=] + |tail|'s [=byte sequence/length=].
+1. If |newSize| &minus; |oldSize| exceeds the available [=storage quota=], throw a {{QuotaExceededError}}.
+1. Set [=this=].[=[[file]]=]'s [=file entry/binary data=] to the concatenation of |head|, |data| and |tail|.
+1. If the operations modifying the [=this=].[=[[file]]=]'s [=file entry/binary data=] in the previous steps
+   failed:
+    1. If there were partial writes and the number of bytes that were written from |data| is known,
+       return the number of bytes that were written from |data|.
+    1. Else return 0.
+1. Return |data|'s [=byte sequence/length=].
+
+</div>
+
+### The {{FileSystemSyncAccessHandle/truncate()}} method ### {#api-filesystemsyncaccesshandle-truncate}
+
+<div class="note domintro">
+  : |handle| . {{FileSystemSyncAccessHandle/truncate()|truncate}}(|newSize|)
+  :: Resizes the file associated with stream to be |newSize| bytes long. If size is larger than the current file size this pads the file with null bytes; otherwise it truncates the file.
+</div>
+
+<div algorithm>
+The <dfn method for=FileSystemSyncAccessHandle>truncate(|newSize|)</dfn> method, when invoked, must run
+these steps:
+
+1. If [=this=].[=[[state]]=] is "`closed`", return [=a promise rejected with=] an {{InvalidStateError}}.
+1. Let |fileContents| be a copy of [=this=].[=[[file]]=]'s [=file entry/binary data=].
+1. Let |p| be [=a new promise=] created in the [=relevant Realm=] of [=this=].
+1. Run the following steps [=in parallel=]:
+  1. Let |oldSize| be the [=byte sequence/length=] of [=this=].[=[[file]]=]'s [=file entry/binary data=].
+  1. If |newSize| is larger than |oldSize|:
+     1. If |newSize| &minus; |oldSize| exceeds the available [=storage quota=], [=/reject=] |p|
+        with a {{QuotaExceededError}} and abort.
+     1. Set [=this=].[=[[file]]=]'s to a [=byte sequence=] formed by concatenating
+        |fileContents| with a [=byte sequence=]
+        containing |newSize| &minus; |oldSize| `0x00` bytes.
+     1. If the operations modifying the [=this=].[=[[file]]=]'s [=file entry/binary data=] in the previous steps
+        failed, [=/reject=] |p| with a {{InvalidStateError}} and abort.
```

Comment:
I agree that most of the actionable error codes would have thrown by now. That being said, do we have to specify what type of error is thrown here? Could this lack of actionability be an argument for throwing the most descriptive error we can? It seems unfortunate to throw all file errors into one bucket. This may not matter in the OPFS, but these file errors are more likely to crop up outside of OPFS, where having a more descriptive error message may help with debugging.

Reply:
After looking at this more closely, returning `InvalidStateError` actually DOES match the current implementation in this case. I agree with Emanuel that we should keep this simple and always throw `InvalidStateError`